ConcurrentModificationException in DefaultPerformanceTracker

Need to investigate cause of this and fix:

```
java.util.concurrent.CompletionException: java.util.ConcurrentModificationException
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:932)
	at java.base/java.util.concurrent.CompletableFuture.uniHandleStage(CompletableFuture.java:946)
	at java.base/java.util.concurrent.CompletableFuture.handle(CompletableFuture.java:2336)
	at tech.pegasys.teku.infrastructure.async.SafeFuture.handle(SafeFuture.java:489)
	at tech.pegasys.teku.infrastructure.async.SafeFuture.finish(SafeFuture.java:301)
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.sendSignedAttestation(ValidatorApiHandler.java:449)
	at jdk.internal.reflect.GeneratedMethodAccessor56.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at tech.pegasys.teku.infrastructure.events.DirectEventDeliverer.executeMethod(DirectEventDeliverer.java:74)
	at tech.pegasys.teku.infrastructure.events.DirectEventDeliverer.deliverTo(DirectEventDeliverer.java:54)
	at tech.pegasys.teku.infrastructure.events.AsyncEventDeliverer.lambda$deliverTo$0(AsyncEventDeliverer.java:61)
	at tech.pegasys.teku.infrastructure.events.AsyncEventDeliverer$QueueReader.deliverNextEvent(AsyncEventDeliverer.java:117)
	at tech.pegasys.teku.infrastructure.events.AsyncEventDeliverer$QueueReader.run(AsyncEventDeliverer.java:109)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:831)
Caused by: java.util.ConcurrentModificationException
	at java.base/java.util.TreeMap.callMappingFunctionWithCheck(TreeMap.java:742)
	at java.base/java.util.TreeMap.computeIfAbsent(TreeMap.java:596)
	at tech.pegasys.teku.validator.coordinator.performance.DefaultPerformanceTracker.saveProducedAttestation(DefaultPerformanceTracker.java:330)
	at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.lambda$sendSignedAttestation$21(ValidatorApiHandler.java:453)
	at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$finish$22(SafeFuture.java:306)
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:930)
	... 16 more
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -27,6 +27,7 @@ For information on changes in released versions of Teku, see the [releases page]
 - Fix an issue where deposits for the PoW chain could be loaded out of order on restart.
 - Fix `/eth/v1/validator/contribution_and_proofs` to return errors.
 - Add `SYNC_COMMITTEE_SUBNET_COUNT` to `/eth/v1/config/spec`, as it was missing.
+- Fixed `ConcurrentModificationException` in validator performance reporting.
 
 ### Experimental: New Altair REST APIs
 - implement POST `/eth/v1/beacon/pool/sync_committees` to allow validators to submit sync committee signatures to the beacon node.

--- a/validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java
+++ b/validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java
@@ -18,6 +18,7 @@ import static com.google.common.base.Preconditions.checkArgument;
 import com.google.common.annotations.VisibleForTesting;
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.IntSummaryStatistics;
@@ -27,6 +28,8 @@ import java.util.NavigableMap;
 import java.util.Optional;
 import java.util.Set;
 import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentSkipListMap;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.stream.Collectors;
@@ -49,11 +52,11 @@ import tech.pegasys.teku.validator.coordinator.ActiveValidatorTracker;
 public class DefaultPerformanceTracker implements PerformanceTracker {
 
   @VisibleForTesting
-  final NavigableMap<UInt64, Set<SlotAndBlockRoot>> producedBlocksByEpoch = new TreeMap<>();
+  final NavigableMap<UInt64, Set<SlotAndBlockRoot>> producedBlocksByEpoch = new ConcurrentSkipListMap<>();
 
-  final NavigableMap<UInt64, Set<Attestation>> producedAttestationsByEpoch = new TreeMap<>();
+  final NavigableMap<UInt64, Set<Attestation>> producedAttestationsByEpoch = new ConcurrentSkipListMap<>();
 
-  final NavigableMap<UInt64, AtomicInteger> blockProductionAttemptsByEpoch = new TreeMap<>();
+  final NavigableMap<UInt64, AtomicInteger> blockProductionAttemptsByEpoch = new ConcurrentSkipListMap<>();
 
   @VisibleForTesting
   static final UInt64 BLOCK_PERFORMANCE_EVALUATION_INTERVAL = UInt64.valueOf(2); // epochs
@@ -225,7 +228,7 @@ public class DefaultPerformanceTracker implements PerformanceTracker {
 
     // Get sent attestations in range
     Set<Attestation> producedAttestations =
-        producedAttestationsByEpoch.getOrDefault(analyzedEpoch, new HashSet<>());
+        producedAttestationsByEpoch.getOrDefault(analyzedEpoch, Collections.emptySet());
     BeaconState state = combinedChainDataClient.getBestState().orElseThrow();
 
     int correctTargetCount = 0;
@@ -327,7 +330,7 @@ public class DefaultPerformanceTracker implements PerformanceTracker {
   public void saveProducedAttestation(Attestation attestation) {
     UInt64 epoch = spec.computeEpochAtSlot(attestation.getData().getSlot());
     Set<Attestation> attestationsInEpoch =
-        producedAttestationsByEpoch.computeIfAbsent(epoch, __ -> new HashSet<>());
+        producedAttestationsByEpoch.computeIfAbsent(epoch, __ -> concurrentSet());
     attestationsInEpoch.add(attestation);
   }
 
@@ -335,7 +338,7 @@ public class DefaultPerformanceTracker implements PerformanceTracker {
   public void saveProducedBlock(SignedBeaconBlock block) {
     UInt64 epoch = spec.computeEpochAtSlot(block.getSlot());
     Set<SlotAndBlockRoot> blocksInEpoch =
-        producedBlocksByEpoch.computeIfAbsent(epoch, __ -> new HashSet<>());
+        producedBlocksByEpoch.computeIfAbsent(epoch, __ -> concurrentSet());
     blocksInEpoch.add(new SlotAndBlockRoot(block.getSlot(), block.getRoot()));
   }
 
@@ -362,5 +365,9 @@ public class DefaultPerformanceTracker implements PerformanceTracker {
 
   static long getPercentage(final long numerator, final long denominator) {
     return (long) (numerator * 100.0 / denominator + 0.5);
+  }
+
+  private <T> Set<T> concurrentSet() {
+    return Collections.newSetFromMap(new ConcurrentHashMap<>());
   }
 }

--- a/validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java
+++ b/validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java
@@ -52,11 +52,14 @@ import tech.pegasys.teku.validator.coordinator.ActiveValidatorTracker;
 public class DefaultPerformanceTracker implements PerformanceTracker {
 
   @VisibleForTesting
-  final NavigableMap<UInt64, Set<SlotAndBlockRoot>> producedBlocksByEpoch = new ConcurrentSkipListMap<>();
+  final NavigableMap<UInt64, Set<SlotAndBlockRoot>> producedBlocksByEpoch =
+      new ConcurrentSkipListMap<>();
 
-  final NavigableMap<UInt64, Set<Attestation>> producedAttestationsByEpoch = new ConcurrentSkipListMap<>();
+  final NavigableMap<UInt64, Set<Attestation>> producedAttestationsByEpoch =
+      new ConcurrentSkipListMap<>();
 
-  final NavigableMap<UInt64, AtomicInteger> blockProductionAttemptsByEpoch = new ConcurrentSkipListMap<>();
+  final NavigableMap<UInt64, AtomicInteger> blockProductionAttemptsByEpoch =
+      new ConcurrentSkipListMap<>();
 
   @VisibleForTesting
   static final UInt64 BLOCK_PERFORMANCE_EVALUATION_INTERVAL = UInt64.valueOf(2); // epochs

